OpenChange/MAPIStoreMailMessage.m

```diff
@@ -311,24 +319,35 @@ - (void) _fetchHeaderData
           if (charset)
             [bodyPartsCharsets setObject: charset forKey: key];
           if (mimeType)
-            [bodyPartsMimeTypes setObject: mimeType forKey: key];
+            {
+              [bodyPartsMimeTypes setObject: mimeType forKey: key];
+              if ([mimeType isEqualToString: @"text/plain"])
+                hasText = YES;
+              else if ([mimeType isEqualToString: @"text/html"])
+                hasHtml = YES;
+            }
 
           if (i == 0)
              {
                ASSIGN (headerMimeType, mimeType);
-               ASSIGN (headerCharset, charset);               
                parameters = partParameters;
              }
-          else
+
+          if (charset)
             {
-              /* We can only put one header charset so we will use utf-8
-                 when we have different charsets in the parts */
-              if (headerCharset != charset)
-                ASSIGN (headerCharset, @"utf-8");
+              if (headerCharset == nil)
+                ASSIGN (headerCharset, charset);
+              else if (![headerCharset isEqualToString: charset])
+                {
+                  /* Because we have different charsets we will encode all in UTF-8 */
+                  ASSIGN (headerCharset, @"utf-8");
+                }
             }
         }
-      
-      
+
+      if (!hasHtml || !hasText)
```

Comment:
I don't understand this. I can have a multipart/mixed with only html or only text